### DISCUSSION TITLE
Borrow notatition from nameless variable binding, while presenting using nameful aproach

### DISCUSSION THREAD

**Discussion by ppolesiuk** [ppolesiuk/type-systems-notes#38] [Ideas]
As pointed out in #28, we can use some notations borrowed from nameless approach of Sec. 1.3, but in "nameful" setting. We can write $\gamma^*$ for simultaneous substitution, $γ \triangleleft_x v$ for extending substitutions, and $γ^{\Uparrow_x}$ to extend substitution $γ$ with dummy mapping for $x$.  Actually we've done it during the last lecture and in the notes in the proof of lambda compatibility (Lemma 4.8 in 16d1e21). I think it is a bit non-standard, but worth considering, since it narrows the gap between formalization and informal reasoning, but it leaves the informal reasoning in a traditional style (with a slightly different notation). If we decide to follow this path, we could add a new section in Chapter 1, that introduces notations used throughout the notes.

What do you think? As this is rather controversial change, I expect some critique (positive or negative) before making any step in that direction.

**Comment by forell** [comment]
I kind of like $γ^*$ for simultaneous substitution. Compared to the common alternatives:
- $γ(e)$ might look a bit "off" when the type of $γ$ is actually Var → Expr, and using $γ^*$ conveys that something more is happening here.
- $e\ γ$ is fine in analogy to $e\\{v/x\\}$, but I don't really see much of a benefit to this one, so might as well be consistent with the nameless notation.

On the other hand, I find $v ◁_x γ$ to be a bit of an eyesore:
- It doesn't follow what I feel is the natural word order here: "extend the substitution $γ$ by a mapping of $x$ to $v$". Trying to fix it by flipping the arrow makes it too similar to the unrelated later operator $▷$.
- It downplays the $x$ for no conceivable reason by putting it in the subscript.

And finally, I think the only place where this notation genuinely appeared was in like 4 subproblems out of 18 in a single exercise a long time ago. It's not defined as a notation in the formalization, it would occasionally get used during the lecture (but not entirely consistently), and it looks quite bad. I'd be happier with something like $γ[x ↦ v]$. I'm probably biased by exposure to it, but I find it easier to understand what's happening at a glance with the more standard notation.

Then there's $γ^{⇑_x}$, which seems kind of superfluous to me, but I'm not inherently opposed to it. What I mean is that in the nameless style, $γ^⇑$ performs a significant operation: it needs to shift the variables in all the values in the substitution. In the named case, we mostly just want to make the substitution an identity on a specific variable (because we're going under a binder), and a whole notation for that seems like a bit much. But maybe it's clearer like this, I don't know. Also, a subscript within a superscript gets really, really tiny! I guess we could typeset it without the subscript, though.

My proposal is to selectively adapt notations from the nameless approach to the named reasoning when they make sense, but without shoehorning in the stuff that doesn't belong. That said, these are all just opinions (and even then most of them focused on just the aesthetics), so I'm open to being convinced otherwise.

**Reply by ppolesiuk** [reply]
For $γ^*$ I agree that it is better than other common alternatives, and it still can be considered standard (the notation is borrowed from bind category theory).

I also agree that the role of $x$ is downplayed in $v ◃_x γ$, and the standard notation $γ[x ↦ v]$ is better. Maybe we should think of $γ[x ↦ v]$ notation analogue in a nameless setting. The natural choice is $γ[v]$, but square brackets are bit overloaded and I'm afraid that it would introduce some confusion. We could also consider $γ[↦v]$.

Finally, the $γ^{⇑_x}$ is an overkill. In named setting it is an identity operation if we assume Barendregt convention. We can omit this operation in pencil-paper proofs or just use non-annotated version $γ^⇑$ to emphasize that the substitution is used under the binder. I don't know which one is better, but I rather prefer skipping this identity operation.

**Reply by forell** [reply]
Yeah, $γ[v]$ seems like a bad idea, but I could live with $γ[↦ v]$ or even $γ[0 ↦ v]$.

**Reply by ppolesiuk** [reply]
Personally, I prefer $γ[↦v]$ over $γ[0↦v]$, because the representation is nameless. The latter suggests modification of the substitution on argument 0 without shifting other mappings, which is not true in this case.
